File: backend/app/schemas/bom.py
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class BomLineIn(BaseModel):
    component_type: str = Field(pattern="^(raw_material|product)$")
    component_id: int
    quantity: float = Field(gt=0)
    unit: str = Field(min_length=1, max_length=20)
    scrap_percent: float = Field(default=0, ge=0, le=100)


class BomLineOut(BaseModel):
    id: int
    parent_product_id: int
    component_type: str
    component_id: int
    component_code: str | None = None
    component_name: str | None = None
    # Set only for component_type == 'raw_material' -- the Raw Material
    # Master's own classification (raw_material/packaging/consumable),
    # so the BOM table can show it without duplicating it as BOM data.
    component_material_type: str | None = None
    # Live stock snapshot for the component (raw materials only) -- read
    # straight from inventory, never stored here. None for a product
    # (sub-assembly) component, which has no meaningful single "on hand"
    # figure the same way (see inventory_service.get_stock's item_type).
    component_on_hand: float | None = None
    quantity: float
    unit: str
    scrap_percent: float
    # quantity inflated by scrap_percent -- the actual amount this line
    # consumes per output_quantity batch. Computed once here (see
    # bom_service._resolve_labels) so every caller (this table, Production,
    # MRP) reads the identical figure instead of each recomputing it.
    effective_quantity: float
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BomReplace(BaseModel):
    """Full replace of a product's BOM: send every line that should exist."""

    lines: list[BomLineIn]

    @field_validator("lines")
    @classmethod
    def _no_duplicate_components(cls, v: list[BomLineIn]) -> list[BomLineIn]:
        seen = set()
        for line in v:
            key = (line.component_type, line.component_id)
            if key in seen:
                raise ValueError(
                    f"Duplicate component in BOM: {line.component_type} #{line.component_id}. "
                    "Combine into a single line instead."
                )
            seen.add(key)
        return v


class RequirementLine(BaseModel):
    """One row of an exploded BOM: how much of this raw material is needed
    in total (across every level of the tree) to build the requested
    quantity of the root product."""

    raw_material_id: int
    code: str | None = None
    name: str | None = None
    unit: str | None = None
    quantity_required: float


class BomExplosionResult(BaseModel):
    product_id: int
    quantity_requested: float
    requirements: list[RequirementLine]


BOM_HEADER_STATUS_PATTERN = "^(active|inactive)$"


class BomHeaderCreate(BaseModel):
    output_quantity: float = Field(default=1, gt=0)
    notes: str | None = Field(default=None, max_length=500)


class BomHeaderUpdate(BaseModel):
    output_quantity: float | None = Field(default=None, gt=0)
    status: str | None = Field(default=None, pattern=BOM_HEADER_STATUS_PATTERN)
    notes: str | None = Field(default=None, max_length=500)


class BomHeaderOut(BaseModel):
    id: int
    bom_number: str
    product_id: int
    output_quantity: float
    status: str
    notes: str | None
    component_count: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/bom.ts
/** Mirrors backend/app/schemas/bom.py. */
import type { RawMaterialType } from './rawMaterial'

export type ComponentType = 'raw_material' | 'product'
export type BomStatus = 'active' | 'inactive'

export interface BomLine {
  id: number
  parent_product_id: number
  component_type: ComponentType
  component_id: number
  component_code: string | null
  component_name: string | null
  /** Only set for component_type === 'raw_material' -- the Raw Material
   * Master's own classification, not BOM data. */
  component_material_type: RawMaterialType | null
  /** Live on-hand stock for the component (raw materials only). */
  component_on_hand: number | null
  quantity: number
  unit: string
  scrap_percent: number
  /** quantity inflated by scrap_percent -- what this line actually
   * consumes per output_quantity batch. Computed server-side so every
   * reader (this table, Production, MRP) sees the identical figure. */
  effective_quantity: number
  created_at: string
  updated_at: string
}

export interface BomLineInput {
  component_type: ComponentType
  component_id: number
  quantity: number
  unit: string
  scrap_percent?: number
}

export interface RequirementLine {
  raw_material_id: number
  code: string | null
  name: string | null
  unit: string | null
  quantity_required: number
}

export interface BomExplosionResult {
  product_id: number
  quantity_requested: number
  requirements: RequirementLine[]
}

/** The BOM "header" -- at most one per product. Owns only batch size,
 * active status and notes; everything about the product itself stays on
 * Product (see backend/app/models/bom.py's Bom docstring). */
export interface BomHeader {
  id: number
  bom_number: string
  product_id: number
  /** The batch size every line's `quantity` is expressed against. */
  output_quantity: number
  status: BomStatus
  notes: string | null
  component_count: number
  created_at: string
  updated_at: string
}

export interface BomHeaderCreateInput {
  output_quantity?: number
  notes?: string | null
}

export interface BomHeaderUpdateInput {
  output_quantity?: number
  status?: BomStatus
  notes?: string | null
}
